### PR TITLE
[16.0][ADD] stock_move_line_change_lot : switch lot reservations between pickings

### PR DESCRIPTION
**Pull request by jbaudoux**
Improve the process of changing a lot on a move line.

Code extracted from shopfloor and refactored in order to have the same behavior in the backend than in shopfloor change lot action.

Changes:
* Never create stock out of nowhere
* Fix available quantity computation and exclude non strictly positive quants
* Better manage lot in package

cc @sebalix @sbejaoui

### CODE REVIEW

**Review by sbejaoui — APPROVED** [review]
Code review + functional test

**Review by lmignon — APPROVED** [review]
LGTM (Code review) + tests

File: stock_move_line_change_lot/models/stock_move_line.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
+# Copyright 2024 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, models
+from odoo.exceptions import UserError
+from odoo.tools.float_utils import float_compare
+
+
+def is_lesser(value, other, rounding):
```
Comment (rousseldenis) [inline stock_move_line_change_lot/models/stock_move_line.py:10]: These functions are great and  more meaningful than float_compare.

They deserve a dedicated module.

And maybe a PR on Odoo core :smiley:

File: stock_move_line_change_lot/models/stock_move_line.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            # unreserve qties of other lines
+            line.unlink()
+            if not is_lesser(freed_quantity, need, rounding):
+                # We reached the required quantity
+                break
+
+        return (freed_quantity, to_reassign_moves)
+
+    def write(self, vals):
+        if self.env.context.get("bypass_reservation_update"):
```
Comment (rousseldenis) [inline stock_move_line_change_lot/models/stock_move_line.py:63]: @jbaudoux IMHO this override is too long. Deserves a separate function and a little bit splitting.

**Review by rousseldenis — APPROVED** [review]
Code review.

Nevertheless, let's move forward on this. Doing some refinements later.

**Pull request merged**